File: pools/allocating_strategies/vlan_strategy.js
// framework managed constants
var currentResources = []
var resourcePoolProperties = {}

// framework managed constants

// STRATEGY_START

/*
VLAN allocation strategy

- Expects VLAN resource type to have 1 properties of type int ["vlan"]
- Logs utilisation stats
- MIN value is 0, MAX value is 4095
- 0 and 4095 are not reserved !
- Allocates previously freed resources
 */

function rangeCapacity(vlanRange) {
    return vlanRange.to - vlanRange.from + 1
}

function freeCapacity(parentRange, utilisedCapacity) {
    return rangeCapacity(parentRange) - utilisedCapacity
}

function utilizedCapacity(allocatedRanges, newlyAllocatedVlan) {
    return allocatedRanges.length + (newlyAllocatedVlan != null)
}

function logStats(newlyAllocatedVlan, parentRange, allocatedVlans = [], level = "log") {
    let utilisedCapacity = utilizedCapacity(allocatedVlans, newlyAllocatedVlan)
    let remainingCapacity = freeCapacity(parentRange, utilisedCapacity)
    let utilPercentage
    if (remainingCapacity === 0) {
        utilPercentage = 100.0
    } else {
        utilPercentage = (utilisedCapacity / rangeCapacity(parentRange)) * 100
    }
    console[level]("Remaining capacity: " + remainingCapacity)
    console[level]("Utilised capacity: " + utilisedCapacity)
    console[level](`Utilisation: ${utilPercentage.toFixed(1)}%`)
}

function invoke() {
    let parentRange = resourcePoolProperties;
    if (parentRange == null) {
        console.error("Unable to allocate VLAN" +
            ". Unable to extract parent vlan range from pool name: " + resourcePoolProperties)
        return null
    }

    // unwrap currentResources
    let currentResourcesUnwrapped = currentResources.map(cR => cR.Properties)
    let currentResourcesSet = new Set(currentResourcesUnwrapped.map(vlan => vlan.vlan))

    for (let i = parentRange.from; i <= parentRange.to; i++) {
        if (!currentResourcesSet.has(i)) {
            // FIXME How to pass these stats ?
            // logStats(i, parentRange, currentResourcesUnwrapped)
            return {
                "vlan": i
            }
        }
    }

    // no more vlans
    console.error("Unable to allocate VLAN from: " + rangeToStr(parentRange) +
        ". Insufficient capacity to allocate a new vlan")
    logStats(null, parentRange, currentResourcesUnwrapped, "error")
    return null
}

function rangeToStr(range) {
    return `[${range.from}-${range.to}]`
}

// STRATEGY_END

// For testing purposes
function invokeWithParams(currentResourcesArg, resourcePoolArg) {
    currentResources = currentResourcesArg
    resourcePoolProperties = resourcePoolArg
    return invoke()
}

exports.invoke = invoke
exports.invokeWithParams = invokeWithParams
exports.utilizedCapacity = utilizedCapacity
exports.freeCapacity = freeCapacity
// For testing purposes

File: pools/allocating_strategies/strategies/go_strategies/vlan/vlan_strategy.go
package vlan

var currentResources []map[string]interface{}
var resourcePoolProperties map[string]interface{}

func utilizedCapacity(allocatedRanges []map[string]interface{}, newlyAllocatedVlan int) int {
	return len(allocatedRanges) + newlyAllocatedVlan
}

func freeCapacity(vlanRange map[string]interface{}, utilisedCapacity int) int {
	return vlanRange["to"].(int) - vlanRange["from"].(int) + 1 - utilisedCapacity
}

func contains(slice []int, val int) bool {
	for _, item := range slice {
		if item == val {
			return true
		}
	}
	return false
}

func invoke() map[string]interface{} {
	if resourcePoolProperties == nil {
		// console.error("Unable to allocate VLAN" +
		// 	". Unable to extract parent vlan range from pool name: " + resourcePoolProperties)
		return nil
	}
	parentRange := make(map[string]interface{})
	for k, v := range resourcePoolProperties {
		parentRange[k] = v
	}

	var currentResourcesUnwrapped []map[string]interface{}
	for _, element := range currentResources {
		value, ok := element["Properties"]
		if ok {
			currentResourcesUnwrapped = append(currentResourcesUnwrapped, value.(map[string]interface{}))
		}
	}
	var currentResourcesSet []int
	for _, element := range currentResourcesUnwrapped {
		value, ok := element["vlan"]
		if ok {
			currentResourcesSet = append(currentResourcesSet, value.(int))
		}
	}
	from, ok := parentRange["from"]
	if !ok {
		return nil
	}
	to, ok := parentRange["to"]
	if !ok {
		return nil
	}
	for i := from.(int); i <= to.(int); i++ {
		if !contains(currentResourcesSet, i) {
			// FIXME How to pass these stats ?
			// logStats(i, parentRange, currentResourcesUnwrapped)
			vlanProperties := make(map[string]interface{})
			vlanProperties["vlan"] = i
			return vlanProperties
		}
	}
	return nil
}

func capacity() map[string]interface{} {
	var result = make(map[string]interface{})
	result["freeCapacity"] = freeCapacity(resourcePoolProperties, len(currentResources))
	result["utilizedCapacity"] = len(currentResources)
	return result
}

// STRATEGY_END

// For testing purposes
func invokeWithParams(
	currentResourcesArg []map[string]interface{},
	resourcePoolArg map[string]interface{}) map[string]interface{} {

	currentResources = currentResourcesArg
	resourcePoolProperties = resourcePoolArg
	return invoke()
}

func invokeWithParamsCapacity(
	currentResourcesArg []map[string]interface{},
	resourcePoolArg map[string]interface{}) map[string]interface{} {

	currentResources = currentResourcesArg
	resourcePoolProperties = resourcePoolArg
	return capacity()
}
